File: sentinelai/hunt.py
"""A small threat-hunting query language over scored windows.

An analyst should not have to edit Python to ask "show me every window with a
high score, few flows and many destinations". This is a deliberately tiny DSL
with the same grammar implemented twice - here for the CLI, and in
web/src/lib/hunt.ts for the console - so a query typed in the browser means
exactly what it means in the terminal.

    probability > 0.9 and distinct_dsts > 20 | sort probability desc | limit 10
    attack ~ brute or failed_logins >= 5
    not entity ~ h00 and night_flag = 1

Grammar (whole thing):

    query    := filter ( '|' directive )*
    filter   := andgroup ( 'or' andgroup )*
    andgroup := clause ( 'and' clause )*
    clause   := [ 'not' ] field op value
    op       := '>' | '>=' | '<' | '<=' | '=' | '==' | '!=' | '~' | '!~'
    value    := number | bareword | quoted string
    directive:= 'sort' field [ 'asc' | 'desc' ] | 'limit' integer

Known limitation, stated rather than hidden: there are no parentheses. `and`
binds tighter than `or`, so `a or b and c` means `a or (b and c)`. Parentheses
were left out because the grammar has to stay small enough to keep the Python and
TypeScript implementations provably identical; a mismatch between them would be
worse than a missing feature.

`~` is a case-insensitive substring match and works on any field, numeric or not.
"""

from __future__ import annotations

import argparse
import json
import re
from dataclasses import dataclass, field as dc_field
from pathlib import Path
from typing import Any, Sequence

OPERATORS = (">=", "<=", "!=", "!~", "==", ">", "<", "=", "~")

_TOKEN = re.compile(
    r"""\s*(?:
        (?P<pipe>\|)
      | (?P<op>>=|<=|!=|!~|==|>|<|=|~)
      | (?P<quoted>'[^']*'|"[^"]*")
      | (?P<word>[A-Za-z_][A-Za-z0-9_.]*)
      | (?P<number>-?\d+(?:\.\d+)?(?:[eE][-+]?\d+)?)
    )""",
    re.VERBOSE,
)


class HuntError(ValueError):
    """Raised for any malformed query. Never swallowed: a query that cannot be
    parsed must fail visibly rather than silently return every row."""


@dataclass(frozen=True)
class Clause:
    field: str
    op: str
    value: float | str
    negated: bool = False

    def matches(self, row: dict[str, Any]) -> bool:
        if self.field not in row:
            raise HuntError(f"unknown field {self.field!r}")
        actual = row[self.field]
        result = _compare(actual, self.op, self.value)
        return not result if self.negated else result


@dataclass(frozen=True)
class Query:
    # OR of AND groups.
    groups: tuple[tuple[Clause, ...], ...] = ()
    sort_field: str | None = None
    sort_desc: bool = True
    limit: int | None = None
    text: str = ""

    def matches(self, row: dict[str, Any]) -> bool:
        if not self.groups:
            return True
        return any(all(c.matches(row) for c in group) for group in self.groups)


def _compare(actual: Any, op: str, expected: float | str) -> bool:
    if op in ("~", "!~"):
        hit = str(expected).lower() in str(actual).lower()
        return hit if op == "~" else not hit

    if isinstance(expected, str):
        # Equality against a string is a plain comparison; ordering is not.
        if op in ("=", "=="):
            return str(actual) == expected
        if op == "!=":
            return str(actual) != expected
        raise HuntError(f"operator {op!r} needs a number, got {expected!r}")

    if actual is None:
        return False
    try:
        left = float(actual)
    except (TypeError, ValueError):
        # Comparing a label against a number is a query bug worth surfacing.
        raise HuntError(f"cannot compare non-numeric value {actual!r} with {op} {expected}")

    if op == ">":
        return left > expected
    if op == ">=":
        return left >= expected
    if op == "<":
        return left < expected
    if op == "<=":
        return left <= expected
    if op in ("=", "=="):
        return left == expected
    if op == "!=":
        return left != expected
    raise HuntError(f"unsupported operator {op!r}")


def _tokenize(query: str) -> list[tuple[str, str]]:
    tokens: list[tuple[str, str]] = []
    pos = 0
    while pos < len(query):
        if query[pos].isspace():
            pos += 1
            continue
        m = _TOKEN.match(query, pos)
        if not m or m.end() == pos:
            raise HuntError(f"cannot parse query at position {pos}: {query[pos:pos + 12]!r}")
        kind = m.lastgroup or ""
        text = (m.group(kind) or "").strip()
        if kind == "quoted":
            text = text[1:-1]
        tokens.append((kind, text))
        pos = m.end()
    return tokens


def parse(query: str, fields: Sequence[str] | None = None) -> Query:
    """Parse a query string. Raises HuntError on anything malformed."""
    tokens = _tokenize(query)
    if not tokens:
        return Query(text=query)

    # Split off the trailing directives first.
    segments: list[list[tuple[str, str]]] = [[]]
    for kind, text in tokens:
        if kind == "pipe":
            segments.append([])
        else:
            segments[-1].append((kind, text))

    groups: list[tuple[Clause, ...]] = []
    current: list[Clause] = []
    body = segments[0]
    i = 0
    negated = False

    while i < len(body):
        kind, text = body[i]
        low = text.lower()

        if kind == "word" and low == "not":
            negated = True
            i += 1
            continue
        if kind == "word" and low == "and":
            i += 1
            continue
        if kind == "word" and low == "or":
            if current:
                groups.append(tuple(current))
                current = []
            i += 1
            continue

        if kind != "word":
            raise HuntError(f"expected a field name, got {text!r}")
        if fields is not None and text not in fields:
            raise HuntError(f"unknown field {text!r}; known fields: {', '.join(sorted(fields))}")
        if i + 2 >= len(body) + 0 or i + 1 >= len(body):
            raise HuntError(f"field {text!r} is missing an operator and value")

        op_kind, op_text = body[i + 1]
        if op_kind != "op":
            raise HuntError(f"expected an operator after {text!r}, got {op_text!r}")
        if i + 2 >= len(body):
            raise HuntError(f"missing a value after {text!r} {op_text}")

        val_kind, val_text = body[i + 2]
        value: float | str
        if val_kind == "number":
            value = float(val_text)
        elif val_kind in ("word", "quoted"):
            value = val_text
        else:
            raise HuntError(f"expected a value after {text!r} {op_text}, got {val_text!r}")

        current.append(Clause(field=text, op=op_text, value=value, negated=negated))
        negated = False
        i += 3

    if negated:
        raise HuntError("trailing 'not' with no clause after it")
    if current:
        groups.append(tuple(current))

    sort_field: str | None = None
    sort_desc = True
    limit: int | None = None

    for segment in segments[1:]:
        if not segment:
            raise HuntError("empty directive after '|'")
        head = segment[0][1].lower()
        if head == "sort":
            if len(segment) < 2:
                raise HuntError("'sort' needs a field name")
            sort_field = segment[1][1]
            if fields is not None and sort_field not in fields:
                raise HuntError(f"cannot sort by unknown field {sort_field!r}")
            if len(segment) > 2:
                direction = segment[2][1].lower()
                if direction not in ("asc", "desc"):
                    raise HuntError(f"sort direction must be asc or desc, got {direction!r}")
                sort_desc = direction == "desc"
        elif head == "limit":
            if len(segment) < 2 or segment[1][0] != "number":
                raise HuntError("'limit' needs an integer")
            limit = int(float(segment[1][1]))
            if limit <= 0:
                raise HuntError("'limit' must be positive")
        else:
            raise HuntError(f"unknown directive {head!r}; expected 'sort' or 'limit'")

    return Query(
        groups=tuple(groups),
        sort_field=sort_field,
        sort_desc=sort_desc,
        limit=limit,
        text=query,
    )


def run(query: str | Query, rows: Sequence[dict[str, Any]], fields: Sequence[str] | None = None
        ) -> list[dict[str, Any]]:
    """Execute a query against rows."""
    q = parse(query, fields) if isinstance(query, str) else query
    hits = [row for row in rows if q.matches(row)]
    if q.sort_field:
        key = q.sort_field

        def sort_key(row: dict[str, Any]) -> tuple[int, float, str]:
            value = row.get(key)
            try:
                return (0, float(value), "")  # type: ignore[arg-type]
            except (TypeError, ValueError):
                return (1, 0.0, str(value))

        hits.sort(key=sort_key, reverse=q.sort_desc)
    if q.limit is not None:
        hits = hits[: q.limit]
    return hits


EXAMPLES: tuple[tuple[str, str], ...] = (
    ("probability > 0.9 | sort probability desc | limit 10", "strongest detections"),
    ("distinct_dsts > 20 and flow_count < 50", "few flows, many hosts: scanning shape"),
    ("failed_logins >= 5 and probability > 0.5", "credential pressure that also scored"),
    ("dns_entropy_max > 3.5", "high-entropy DNS: tunnelling shape"),
    ("attack ~ brute and probability < 0.63", "missed brute force: below threshold"),
    ("graph_score > 0 and night_flag = 1", "auth-graph movement out of hours"),
    ("night_flag = 1 and bytes_out_sum > 1e6", "large egress overnight"),
)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Hunt over scored windows")
    ap.add_argument("query", nargs="?", help="query string; omit to list examples")
    ap.add_argument("--artifacts", default="artifacts")
    ap.add_argument("--columns", default="entity,win,probability,attack")
    args = ap.parse_args(argv)

    index_path = Path(args.artifacts) / "hunt_index.json"
    if not index_path.exists():
        raise SystemExit(
            f"missing {index_path}\nRun: python -m sentinelai.entity_risk --artifacts {args.artifacts}"
        )
    index = json.loads(index_path.read_text(encoding="utf-8"))
    rows: list[dict[str, Any]] = index["rows"]
    fields: list[str] = index["fields"]

    if not args.query:
        print(f"{len(rows)} windows indexed. Fields:")
        print("  " + ", ".join(fields))
        print("\nExamples:")
        for q, why in EXAMPLES:
            print(f"  {q}\n      {why}")
        return 0

    try:
        hits = run(args.query, rows, fields)
    except HuntError as exc:
        raise SystemExit(f"bad query: {exc}")

    columns = [c for c in args.columns.split(",") if c in fields]
    print(f"{len(hits)} of {len(rows)} windows match\n")
    print("  ".join(f"{c:<14}" for c in columns))
    for row in hits[:50]:
        cells = []
        for c in columns:
            v = row.get(c)
            cells.append(f"{v:<14.6f}" if isinstance(v, float) else f"{str(v):<14}")
        print("  ".join(cells))
    if len(hits) > 50:
        print(f"... {len(hits) - 50} more")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: web/src/lib/hunt.ts
/**
 * The hunting query language, in the browser.
 *
 * This is a deliberate second implementation of sentinelai/hunt.py. The grammar
 * is small precisely so the two can stay identical: a query typed into the
 * console means exactly what the same query means in the CLI. Every operator,
 * precedence rule and error case below mirrors the Python module.
 *
 *   probability > 0.9 and distinct_dsts > 20 | sort probability desc | limit 10
 *   attack ~ brute or failed_logins >= 5
 *   not entity ~ h00 and night_flag = 1
 *
 * Known limitation, same as the CLI: no parentheses. `and` binds tighter than
 * `or`, so `a or b and c` parses as `a or (b and c)`.
 */

export type HuntRow = Record<string, string | number | null>

export type Clause = {
	field: string
	op: string
	value: number | string
	negated: boolean
}

export type Query = {
	/** OR of AND groups. */
	groups: Clause[][]
	sortField: string | null
	sortDesc: boolean
	limit: number | null
	text: string
}

/** Thrown for any malformed query. Never swallowed: a broken query must not
 *  silently return every row. */
export class HuntError extends Error {}

type Token = { kind: 'pipe' | 'op' | 'word' | 'number' | 'quoted'; text: string }

const OP_RE = /^(>=|<=|!=|!~|==|>|<|=|~)/
const WORD_RE = /^[A-Za-z_][A-Za-z0-9_.]*/
const NUMBER_RE = /^-?\d+(?:\.\d+)?(?:[eE][-+]?\d+)?/

function tokenize(input: string): Token[] {
	const tokens: Token[] = []
	let rest = input
	let pos = 0
	while (rest.length > 0) {
		const ws = rest.match(/^\s+/)
		if (ws) {
			const len = ws[0].length
			rest = rest.slice(len)
			pos += len
			continue
		}
		if (rest.startsWith('|')) {
			tokens.push({ kind: 'pipe', text: '|' })
			rest = rest.slice(1)
			pos += 1
			continue
		}
		// Operators are matched before words so `~` never becomes part of a name.
		const op = rest.match(OP_RE)
		if (op) {
			tokens.push({ kind: 'op', text: op[0] })
			rest = rest.slice(op[0].length)
			pos += op[0].length
			continue
		}
		const quote = rest[0]
		if (quote === "'" || quote === '"') {
			const end = rest.indexOf(quote, 1)
			if (end < 0) throw new HuntError(`unterminated string at position ${pos}`)
			tokens.push({ kind: 'quoted', text: rest.slice(1, end) })
			rest = rest.slice(end + 1)
			pos += end + 1
			continue
		}
		// Numbers before words, so -1.5e3 is not split apart.
		const num = rest.match(NUMBER_RE)
		if (num) {
			tokens.push({ kind: 'number', text: num[0] })
			rest = rest.slice(num[0].length)
			pos += num[0].length
			continue
		}
		const word = rest.match(WORD_RE)
		if (word) {
			tokens.push({ kind: 'word', text: word[0] })
			rest = rest.slice(word[0].length)
			pos += word[0].length
			continue
		}
		throw new HuntError(`cannot parse query at position ${pos}: ${rest.slice(0, 12)}`)
	}
	return tokens
}

export function parse(input: string, fields?: readonly string[]): Query {
	const tokens = tokenize(input)
	const empty: Query = { groups: [], sortField: null, sortDesc: true, limit: null, text: input }
	if (tokens.length === 0) return empty

	const segments: Token[][] = [[]]
	for (const token of tokens) {
		if (token.kind === 'pipe') segments.push([])
		else segments[segments.length - 1]!.push(token)
	}

	const body = segments[0] ?? []
	const groups: Clause[][] = []
	let current: Clause[] = []
	let negated = false
	let i = 0

	while (i < body.length) {
		const token = body[i]!
		const low = token.text.toLowerCase()

		if (token.kind === 'word' && low === 'not') {
			negated = true
			i += 1
			continue
		}
		if (token.kind === 'word' && low === 'and') {
			i += 1
			continue
		}
		if (token.kind === 'word' && low === 'or') {
			if (current.length > 0) {
				groups.push(current)
				current = []
			}
			i += 1
			continue
		}

		if (token.kind !== 'word') {
			throw new HuntError(`expected a field name, got ${token.text}`)
		}
		if (fields && !fields.includes(token.text)) {
			throw new HuntError(`unknown field ${token.text}`)
		}
		const opToken = body[i + 1]
		if (!opToken || opToken.kind !== 'op') {
			throw new HuntError(`expected an operator after ${token.text}`)
		}
		const valToken = body[i + 2]
		if (!valToken) {
			throw new HuntError(`missing a value after ${token.text} ${opToken.text}`)
		}
		const value: number | string =
			valToken.kind === 'number' ? Number(valToken.text) : valToken.text
		if (valToken.kind === 'op' || valToken.kind === 'pipe') {
			throw new HuntError(`expected a value after ${token.text} ${opToken.text}`)
		}

		current.push({ field: token.text, op: opToken.text, value, negated })
		negated = false
		i += 3
	}

	if (negated) throw new HuntError("trailing 'not' with no clause after it")
	if (current.length > 0) groups.push(current)

	let sortField: string | null = null
	let sortDesc = true
	let limit: number | null = null

	for (const segment of segments.slice(1)) {
		const head = segment[0]
		if (!head) throw new HuntError("empty directive after '|'")
		const name = head.text.toLowerCase()
		if (name === 'sort') {
			const target = segment[1]
			if (!target) throw new HuntError("'sort' needs a field name")
			if (fields && !fields.includes(target.text)) {
				throw new HuntError(`cannot sort by unknown field ${target.text}`)
			}
			sortField = target.text
			const dir = segment[2]
			if (dir) {
				const d = dir.text.toLowerCase()
				if (d !== 'asc' && d !== 'desc') {
					throw new HuntError(`sort direction must be asc or desc, got ${dir.text}`)
				}
				sortDesc = d === 'desc'
			}
		} else if (name === 'limit') {
			const n = segment[1]
			if (!n || n.kind !== 'number') throw new HuntError("'limit' needs an integer")
			limit = Math.trunc(Number(n.text))
			if (limit <= 0) throw new HuntError("'limit' must be positive")
		} else {
			throw new HuntError(`unknown directive ${name}; expected 'sort' or 'limit'`)
		}
	}

	return { groups, sortField, sortDesc, limit, text: input }
}

function compare(actual: unknown, op: string, expected: number | string): boolean {
	if (op === '~' || op === '!~') {
		const hit = String(actual).toLowerCase().includes(String(expected).toLowerCase())
		return op === '~' ? hit : !hit
	}
	if (typeof expected === 'string') {
		if (op === '=' || op === '==') return String(actual) === expected
		if (op === '!=') return String(actual) !== expected
		throw new HuntError(`operator ${op} needs a number, got ${expected}`)
	}
	if (actual === null || actual === undefined) return false
	const left = Number(actual)
	if (Number.isNaN(left)) {
		throw new HuntError(`cannot compare non-numeric value ${String(actual)} with ${op}`)
	}
	switch (op) {
		case '>':
			return left > expected
		case '>=':
			return left >= expected
		case '<':
			return left < expected
		case '<=':
			return left <= expected
		case '=':
		case '==':
			return left === expected
		case '!=':
			return left !== expected
		default:
			throw new HuntError(`unsupported operator ${op}`)
	}
}

function matches(query: Query, row: HuntRow): boolean {
	if (query.groups.length === 0) return true
	return query.groups.some((group) =>
		group.every((clause) => {
			if (!(clause.field in row)) throw new HuntError(`unknown field ${clause.field}`)
			const hit = compare(row[clause.field], clause.op, clause.value)
			return clause.negated ? !hit : hit
		}),
	)
}

export function run(
	input: string | Query,
	rows: readonly HuntRow[],
	fields?: readonly string[],
): HuntRow[] {
	const query = typeof input === 'string' ? parse(input, fields) : input
	const hits = rows.filter((row) => matches(query, row))
	if (query.sortField) {
		const key = query.sortField
		hits.sort((a, b) => {
			const av = Number(a[key])
			const bv = Number(b[key])
			if (Number.isNaN(av) || Number.isNaN(bv)) {
				return String(a[key]).localeCompare(String(b[key])) * (query.sortDesc ? -1 : 1)
			}
			return query.sortDesc ? bv - av : av - bv
		})
	}
	return query.limit === null ? hits : hits.slice(0, query.limit)
}

/** The same examples the CLI prints, so both surfaces teach the same syntax. */
export const HUNT_EXAMPLES: ReadonlyArray<{ query: string; why: string }> = [
	{ query: 'probability > 0.9 | sort probability desc | limit 10', why: 'strongest detections' },
	{ query: 'distinct_dsts > 20 and flow_count < 50', why: 'few flows, many hosts: scanning' },
	{ query: 'failed_logins >= 5 and probability > 0.5', why: 'credential pressure that scored' },
	{ query: 'dns_entropy_max > 3.5', why: 'high-entropy DNS: tunnelling' },
	{ query: 'attack ~ brute and probability < 0.63', why: 'missed brute force: below threshold' },
	{ query: 'graph_score > 0 and night_flag = 1', why: 'auth-graph movement out of hours' },
	{ query: 'night_flag = 1 and bytes_out_sum > 1e6', why: 'large egress overnight' },
]